### PR TITLE
Avoid name conflict by using deployment+delimiter+container as the key

### LINKED ISSUE #148: Improve the container image override mechanism
**Problem**
As seen in https://github.com/knative/eventing/issues/2794, we cannot simply use the container names for overriding images for containers. Container names can be duplicated.

Current override mechanism is like this:

```
apiVersion: operator.knative.dev/v1alpha1
kind: KnativeEventing
metadata:
...
spec:
  registry:
    override:
      eventing-controller: docker.io/my-eventing-controller
      eventing-webhook: docker.io/my-eventing-webhook
```

This cannot handle the case that we have for `imc-controller` and `broker-controller` deployments the same container name, `controller`.

My suggestion would be instead of  only providing the container name, we pass more information. Something like:

```
apiVersion: operator.knative.dev/v1alpha1
kind: KnativeEventing
metadata:
...
spec:
  registry:
    overrides:
    - namespace: knative-eventing
      deployment: eventing-controller
      container: eventing-controller
      image: docker.io/my-eventing-controller
    - namespace: knative-eventing
      deployment: eventing-webhook
      container: eventing-webhook
      image: docker.io/my-eventing-webhook
```

This would fix the case that we have for `imc-controller` and `broker-controller` deployments the same container name, `controller`.

**[Persona:](https://github.com/knative/eventing/blob/master/docs/personas.md)**
Which persona is this feature for?
System admin

**Exit Criteria**
Proper image overriding, regardless of container names being unique or not.

**Time Estimate (optional):**
1 day

**Additional context (optional)**

See the causing ticket: https://github.com/knative/eventing/issues/2794

Related: https://github.com/knative/eventing-operator/issues/147

### PR DESCRIPTION
**Pull request by houshengbo**
# Issue to be fixed

Fixes #148

## Proposed Changes

*
*
*

## Release Note

```release-note
```

### CODE REVIEW

File: pkg/reconciler/knativeeventing/common/images.go [hunk 1]
```diff
@@ -33,6 +33,7 @@ import (
 var (
 	// The string to be replaced by the container name
 	containerNameVariable = "${NAME}"
+	delimiter = "-"
```
Comment (mattmoor) [inline pkg/reconciler/knativeeventing/common/images.go]: Format Go code:
```suggestion
	delimiter             = "-"
```

File: pkg/reconciler/knativeeventing/common/images.go [hunk 2]
```diff
@@ -92,16 +93,19 @@ func updateDeploymentImage(deployment *appsv1.Deployment, registry *eventingv1al
...
 			updateContainer(container, newImage, log)
 		}
 	}
 	log.Debugw("Finished updating images", "name", deployment.GetName(), "containers", deployment.Spec.Template.Spec.Containers)
 }
 
-func getNewImage(registry *eventingv1alpha1.Registry, containerName string) string {
-	overrideImage := registry.Override[containerName]
+func getNewImage(registry *eventingv1alpha1.Registry, containerName, deploymentName string) string {
+	overrideImage := registry.Override[deploymentName + delimiter + containerName]
```
Comment (mattmoor) [inline pkg/reconciler/knativeeventing/common/images.go]: Format Go code:
```suggestion
	overrideImage := registry.Override[deploymentName+delimiter+containerName]
```

File: pkg/reconciler/knativeeventing/common/images.go (hunk 1 above)
Comment (mattmoor) [inline pkg/reconciler/knativeeventing/common/images.go]: Format Go code:
```suggestion
	delimiter             = "-"
```

File: pkg/reconciler/knativeeventing/common/images.go [hunk 3]
```diff
@@ -33,6 +33,7 @@ import (
 var (
 	// The string to be replaced by the container name
 	containerNameVariable = "${NAME}"
+	delimiter             = "-"
```
Comment (Cynocracy) [inline pkg/reconciler/knativeeventing/common/images.go]: I thought the decision from chat was to use / as a delimiter, am I misremembering?

In particular, because - is allowed in deployment's container names, the override for

deployment-container

is ambiguous, it could mean container in deployment, or it could mean a container called deployment-container.

File: pkg/reconciler/knativeeventing/common/images.go [hunk 4]
```diff
@@ -92,16 +93,19 @@ func updateDeploymentImage(deployment *appsv1.Deployment, registry *eventingv1al
...
 			updateContainer(container, newImage, log)
 		}
 	}
 	log.Debugw("Finished updating images", "name", deployment.GetName(), "containers", deployment.Spec.Template.Spec.Containers)
 }
 
-func getNewImage(registry *eventingv1alpha1.Registry, containerName string) string {
-	overrideImage := registry.Override[containerName]
+func getNewImage(registry *eventingv1alpha1.Registry, containerName, deploymentName string) string {
+	overrideImage := registry.Override[deploymentName+delimiter+containerName]
```
Comment (Cynocracy) [inline pkg/reconciler/knativeeventing/common/images.go]: Re-using the same override map for deployment and deployment-container style overrides could cause issues if the deployment name contains -, which seems to be possible.

Thoughts on creating two maps, one for container-only overrides and one for deployment+container?

File: pkg/reconciler/knativeeventing/common/images_test.go [hunk 5]
```diff
@@ -172,6 +172,96 @@ var updateDeploymentImageTests = []updateDeploymentImageTest{
 			Env:   []corev1.EnvVar{{Name: "SOME_IMAGE", Value: "docker.io/my/overridden-image"}},
 		}},
 	},
+	{
```
Comment (Cynocracy) [inline pkg/reconciler/knativeeventing/common/images_test.go]: It would be nice to test a few more corner cases:

* container X in multiple deployments
* Specify just X-Y, doesn't apply to container called X-Y

File: pkg/reconciler/knativeeventing/common/images.go (hunk 3 above)
Comment (houshengbo) [reply pkg/reconciler/knativeeventing/common/images.go]: It looks like to I need to switch to `/`.

File: pkg/reconciler/knativeeventing/common/images.go (hunk 4 above)
Comment (houshengbo) [reply pkg/reconciler/knativeeventing/common/images.go]: I have changed the delimiter to `/`.

**Comment by aliok** [comment]
I wanted to give this a try today.
All works fine. Precedence of deployment+containerName over containerName is also good.
Image overrides with env vars continue working fine, no regressions there.

Here are the cases I tried:

```
cat <<EOS |kubectl apply -f -
---
apiVersion: operator.knative.dev/v1alpha1
kind: KnativeEventing
metadata:
  name: knative-eventing
  namespace: knative-eventing
spec:
  registry:
    override:
      eventing-controller: docker.io/foo
EOS
```

```
cat <<EOS |kubectl apply -f -
---
apiVersion: operator.knative.dev/v1alpha1
kind: KnativeEventing
metadata:
  name: knative-eventing
  namespace: knative-eventing
spec:
  registry:
    override:
      eventing-controller/eventing-controller: docker.io/foo
EOS
```

```
cat <<EOS |kubectl apply -f -
---
apiVersion: operator.knative.dev/v1alpha1
kind: KnativeEventing
metadata:
  name: knative-eventing
  namespace: knative-eventing
spec:
  registry:
    override:
      eventing-controller: docker.io/foo1
      eventing-controller/eventing-controller: docker.io/foo2
EOS
```

```
cat <<EOS |kubectl apply -f -
---
apiVersion: operator.knative.dev/v1alpha1
kind: KnativeEventing
metadata:
  name: knative-eventing
  namespace: knative-eventing
spec:
  registry:
    override:
      eventing-controller: docker.io/foo1
      eventing-controller/eventing-controller: docker.io/foo2
      PING_IMAGE: docker.io/foo_ping
EOS
```
